src/openloco/windows/error.cpp

```diff
@@ -0,0 +1,305 @@
+#include "../audio/audio.h"
+#include "../companymgr.h"
+#include "../graphics/colours.h"
+#include "../graphics/image_ids.h"
+#include "../interop/interop.hpp"
+#include "../localisation/FormatArguments.hpp"
+#include "../localisation/stringmgr.h"
+#include "../objects/competitor_object.h"
+#include "../objects/interface_skin_object.h"
+#include "../objects/objectmgr.h"
+#include "../openloco.h"
+#include "../ui/WindowManager.h"
+
+using namespace openloco::interop;
+
+namespace openloco::ui::windows::error
+{
+    static loco_global<uint8_t, 0x00508F09> _suppressErrorSound;
+    static loco_global<ui::widget_t[1], 0x00508F1C> _widgets;
+    static loco_global<char[512], 0x009C64B4> _byte_9C64B3;
+    static loco_global<uint16_t, 0x009C66B3> _word_9C66B3;
+    static loco_global<uint8_t, 0x009C68EC> _byte_9C68EC;
+    static loco_global<int32_t, 0x112C876> gCurrentFontSpriteBase;
+    static loco_global<int32_t, 0x0113E72C> _cursorX;
+    static loco_global<int32_t, 0x0113E730> _cursorY;
+
+    namespace common
+    {
+        static window_event_list events;
+
+        static void draw(ui::window* self, gfx::drawpixelinfo_t* dpi);
+        static void on_periodic_update(ui::window* self);
+        static void init_events();
+    }
+    namespace error
+    {
+        enum widx
+        {
+            frame,
+        };
+
+        widget_t widgets[] = {
+            make_widget({ 0, 0 }, { 200, 42 }, widget_type::wt_3, 0),
+            widget_end(),
+        };
+    }
+
+    namespace errorCompetitor
+    {
+        enum widx
+        {
+            frame,
+            innerFrame,
+        };
+
+        widget_t widgets[] = {
+            make_widget({ 0, 0 }, { 250, 70 }, widget_type::wt_3, 0),
+            make_widget({ 3, 3 }, { 64, 64 }, widget_type::wt_3, 1),
+            widget_end(),
+        };
+    }
+
+    static char* formatErrorString(string_id title, string_id message, char* buffer)
+    {
+        char* ptr = (char*)buffer;
+        ptr[0] = -112;
+        ptr++;
+
+        if (title != string_ids::null)
+        {
+            auto args = FormatArguments();
+
+            ptr = stringmgr::format_string(ptr, title, &args);
+        }
+
+        if (message != string_ids::null)
+        {
+            if (title != string_ids::null)
+            {
+
+                *ptr = control_codes::newline;
+                ptr++;
+            }
+
+            auto args = FormatArguments();
+
+            stringmgr::format_string(ptr, message, &args);
+        }
+        return ptr;
+    }
+
+    // 0x00431A8A
+    void open(string_id title, string_id message)
+    {
+        _byte_9C68EC = 0xFF;
+
+        WindowManager::close(WindowType::error);
+
+        char* buffer = _byte_9C64B3;
+
+        buffer = formatErrorString(title, message, buffer);
+
+        if (buffer[0] != _byte_9C64B3[0])
```

Comment:
this is an address comparison not a value comparison. We are checking to see if formatErrorString actually did anything and there is a string buffer to use.